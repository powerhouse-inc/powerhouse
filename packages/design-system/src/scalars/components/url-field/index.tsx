export { UrlField, type UrlFieldProps } from "./url-field.js";
export {
  UrlInput,
  type PlatformIcon,
  type UrlInputProps,
} from "./url-input.js";
