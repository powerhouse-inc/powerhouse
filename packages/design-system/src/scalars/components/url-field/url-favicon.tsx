import { iconNames, type IconName } from "#powerhouse";
import { Icon } from "#powerhouse";
import { useMemo } from "react";

interface UrlFaviconProps {
  url: string;
  platformIcons?: Record<string, React.ReactElement | IconName>;
}

const DEFAULT_ICON_NAME = "GlobeWww";

const UrlFavicon = ({ url, platformIcons }: UrlFaviconProps) => {
  const IconToUse: React.ReactElement | null = useMemo(() => {
    const defaultIcon = <Icon name={DEFAULT_ICON_NAME} size={18} />;
    try {
      const parsedUrl = new URL(url);
      const hostname = parsedUrl.hostname;
      const icon = platformIcons?.[hostname];
      if (!icon) return defaultIcon;

      if (typeof icon === "string") {
        // check beyond typescript check if the icon actually exists
        if (iconNames.includes(icon)) {
          return <Icon name={icon} size={18} />;
        } else {
          console.error(`Platform icon "${icon}" does not exist`);
          return defaultIcon;
        }
      }

      return icon;
    } catch {
      return defaultIcon;
    }
  }, [url, platformIcons]);

  if (Object.keys(platformIcons ?? {}).length === 0) return null;

  return (
    <div className="absolute left-2.5 top-0 flex h-full items-center justify-center text-gray-900">
      {IconToUse}
    </div>
  );
};

export default UrlFavicon;
