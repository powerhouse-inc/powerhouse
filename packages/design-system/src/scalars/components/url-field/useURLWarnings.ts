import { useCallback, useEffect, useState } from "react";

export const useURLWarnings = (value: string) => {
  const [warnings, setWarnings] = useState<string[]>([]);

  const checkForWarnings = useCallback(() => {
    const detectedWarnings: string[] = [];

    try {
      const url = new URL(value);
      // check for unencoded spaces in the URL
      if (value.trim().includes(" ")) {
        detectedWarnings.push("Using unencoded spaces in the URL");
      }
      // check for double slashes in the URL
      if (value.indexOf("//") !== value.lastIndexOf("//")) {
        detectedWarnings.push("Using double slashes in the URL");
      }
      // check for backslashes in the URL
      if (value.includes("\\")) {
        detectedWarnings.push("Using backslashes in the URL");
      }
      // using ? character twice in the URL
      if (value.indexOf("?") !== value.lastIndexOf("?")) {
        detectedWarnings.push("Using ? character twice in the URL");
      }
      // using # character twice in the URL
      if (value.indexOf("#") !== value.lastIndexOf("#")) {
        detectedWarnings.push("Using # character twice in the URL");
      }
      // end with ... maybe it was truncated
      if (value.endsWith("...")) {
        detectedWarnings.push("URL may be truncated");
      }

      // check for incorrect tld
      const tld = url.hostname.split(".").pop()!;
      if (
        [
          "con",
          "comm",
          "comn",
          "cmo",
          "copm",
          "om",
          "ccom",
          "con",
          "nett",
          "orgg",
        ].includes(tld)
      ) {
        detectedWarnings.push("URL may have incorrect top-level domain (TLD)");
      }

      // the user probably forgot the ".""
      if (url.hostname.endsWith("com") && tld.length > 3) {
        detectedWarnings.push("URL may be missing the '.' before the .com");
      }

      setWarnings(detectedWarnings);
    } catch {
      // the url is not valid so we don't check for warnings
    }
  }, [value]);

  // reset warnings when the value is empty or not a valid url
  useEffect(() => {
    if (warnings.length === 0) return;

    if (value === "") {
      setWarnings([]);
    }

    try {
      new URL(value);
    } catch {
      setWarnings([]);
    }
  }, [value, checkForWarnings]);

  return {
    warnings,
    checkForWarnings,
  };
};
