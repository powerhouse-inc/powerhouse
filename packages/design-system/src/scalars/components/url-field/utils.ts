import { IconName } from "@/assets/icon-components/types";

export function getIconName(url: string): IconName {
  const defaultIcon = "GlobeWww";

  if (!url.trim()) return defaultIcon;

  try {
    const parsedUrl = new URL(url);
    const hostname = parsedUrl.hostname;
    switch (hostname) {
      case "forum.sky.money":
        return "Forum";
      case "discord.com":
      case "discord.gg":
        return "Discord";
      case "twitter.com":
      case "x.com":
        return "XTwitter";
      case "github.com":
        return "Github";
      case "linkedin.com":
        return "Linkedin";
      case "youtube.com":
      case "youtu.be":
        return "Youtube";
      default:
        return defaultIcon;
    }
  } catch {
    // If the URL is invalid, return the default icon
    return defaultIcon;
  }
}
