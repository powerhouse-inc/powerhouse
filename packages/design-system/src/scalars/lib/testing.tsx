import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
