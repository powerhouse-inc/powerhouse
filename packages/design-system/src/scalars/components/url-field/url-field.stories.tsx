import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { UrlField } from "./url-field.js";

const meta: Meta<typeof UrlField> = {
  title: "Document Engineering/Scalars/Url Field",
  component: UrlField,
  decorators: [withForm],
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,

    allowedProtocols: {
      control: "object",
      description: "Allowed protocols for the URL",
      table: {
        type: { summary: "Array<string>" },
        defaultValue: { summary: "['http', 'https']" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    maxURLLength: {
      control: "number",
      description: "Maximum length of the URL",
      table: {
        type: { summary: "number" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    platformIcons: {
      control: "object",
      description:
        "An object where the key is the hostname and the value is the icon name or a React element",
      table: {
        type: { summary: "Record<string, IconName | ReactElement>" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...getValidationArgTypes(),
  },
  args: {
    name: "url-field",
    allowedProtocols: ["https"],
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: "Website URL",
    placeholder: "https://example.com",
  },
};

export const WithDescription: Story = {
  args: {
    label: "Forum URL",
    description: "Enter the URL to your online forum",
    placeholder: "https://myforum.com",
  },
};

export const Required: Story = {
  args: {
    label: "LinkedIn",
    required: true,
    placeholder: "https://linkedin.com/in/username",
  },
};

export const WithValue: Story = {
  args: {
    label: "With Value",
    value: "https://exaomple.com",
  },
};

export const Disabled: Story = {
  args: {
    label: "Read Only URL",
    value: "https://example.com",
    disabled: true,
  },
};

export const WithError: Story = {
  args: {
    label: "Website URL",
    value: "not-a-valid-url",
    errors: ["Please enter a valid URL"],
  },
};

export const WithWarning: Story = {
  args: {
    label: "Website URL",
    value: "https://example.com",
    warnings: ["URL may be unreachable"],
  },
};

export const WithPlatformIcon: Story = {
  args: {
    label: "Website URL",
    value: "https://github.com/test",
    platformIcons: {
      "github.com": "Github",
      "linkedin.com": "Linkedin",
      "twitter.com": "XTwitter",
      "x.com": "XTwitter",
      "youtube.com": "Youtube",
      "forum.sky.money": "Forum",
      "discord.com": "Discord",
      "discord.gg": "Discord",
    },
  },
};
