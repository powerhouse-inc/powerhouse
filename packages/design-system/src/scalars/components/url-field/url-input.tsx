import { type IconName } from "#powerhouse";
import React, { useCallback, useId, useMemo } from "react";
import { Input } from "../../../ui/components/index.js";
import { sharedValueTransformers } from "../../lib/shared-value-transformers.js";
import { cn } from "../../lib/utils.js";
import {
  FormDescription,
  FormGroup,
  FormLabel,
  FormMessageList,
} from "../fragments/index.js";
import ValueTransformer from "../fragments/value-transformer/index.js";
import type { InputBaseProps } from "../types.js";
import UrlFavicon from "./url-favicon.js";
import { useURLWarnings } from "./useURLWarnings.js";

type PlatformIcon = IconName | React.ReactElement;

interface UrlInputProps
  extends InputBaseProps<string>,
    // FieldErrorHandling,
    Omit<
      React.InputHTMLAttributes<HTMLInputElement>,
      "pattern" | "value" | "defaultValue" | "name" | "maxLength"
    > {
  showWarnings?: boolean;
  platformIcons?: Record<string, PlatformIcon>;
}

const UrlInput = React.forwardRef<HTMLInputElement, UrlInputProps>(
  (
    {
      label,
      description,
      showWarnings = true,
      warnings: warningsProp,
      errors,
      platformIcons,
      value,
      onBlur,

      ...props
    },
    ref,
  ) => {
    const idGenerated = useId();
    const id = props.id ?? idGenerated;
    const hasError = !!errors?.length;
    const { warnings, checkForWarnings } = useURLWarnings(value ?? "");
    const showIcon = Object.keys(platformIcons ?? {}).length > 0;

    const combinedWarnings = useMemo(() => {
      return [...(warningsProp ?? []), ...warnings];
    }, [warningsProp, warnings]);

    const handleBlur = useCallback(
      (event: React.FocusEvent<HTMLInputElement>) => {
        checkForWarnings();
        onBlur?.(event);
      },
      [checkForWarnings, onBlur],
    );

    const handleWarningsOnEnter = useCallback(
      (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter") {
          checkForWarnings();
        }
      },
      [checkForWarnings],
    );

    // prevent url from having trailing spaces
    const transformers = useMemo(
      () => [sharedValueTransformers.trimOnBlur()],
      [],
    );

    return (
      <FormGroup>
        <FormLabel
          htmlFor={id}
          required={props.required}
          disabled={props.disabled}
          hasError={!!errors?.length}
        >
          {label}
        </FormLabel>
        <div className="relative">
          <ValueTransformer transformers={transformers}>
            <Input
              id={id}
              ref={ref}
              type="url"
              {...props}
              value={value ?? ""}
              onBlur={handleBlur}
              onKeyDown={handleWarningsOnEnter}
              aria-invalid={hasError}
              className={cn(showIcon && "pl-8")}
              data-cast="URLTrim"
            />
          </ValueTransformer>
          <UrlFavicon url={value ?? ""} platformIcons={platformIcons} />
        </div>
        {description && <FormDescription>{description}</FormDescription>}
        {showWarnings && combinedWarnings.length > 0 && (
          <FormMessageList messages={combinedWarnings} type="warning" />
        )}
        {errors && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    );
  },
);

UrlInput.displayName = "UrlInput";

export { UrlInput, type PlatformIcon, type UrlInputProps };
