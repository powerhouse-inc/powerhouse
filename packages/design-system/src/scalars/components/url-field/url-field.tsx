import { withFieldValidation } from "../fragments/with-field-validation/with-field-validation.js";
import type { FieldErrorHandling } from "../types.js";
import { UrlInput, type UrlInputProps } from "./url-input.js";

interface UrlFieldProps extends UrlInputProps, FieldErrorHandling {
  allowedProtocols?: string[];
  maxURLLength?: number;
}

const UrlField = withFieldValidation<UrlFieldProps>(UrlInput, {
  validations: {
    _validUrl: () => (value) => {
      if (!value) return true;
      try {
        const url = new URL(value as string);
        const regex =
          /^(([0-9]{1,3}\.){3}[0-9]{1,3})|(([a-z0-9-_]+\.)+[a-z]{2,})|(localhost)$/i;
        if (!regex.test(url.hostname)) {
          return `${value} must be a valid URL`;
        }
        return true;
      } catch {
        return `${value} must be a valid URL`;
      }
    },
    _allowedProtocols:
      ({ allowedProtocols }) =>
      (value) => {
        if (!value || !allowedProtocols) return true;
        try {
          const url = new URL(value as string);
          const isAllowed = allowedProtocols.includes(
            url.protocol.slice(0, -1),
          );
          if (isAllowed) return true;
        } catch {
          return true; // handling valid url is handled by _validUrl validation
        }

        let allowedProtocolsString = allowedProtocols.join(", ");
        if (allowedProtocols.length > 1) {
          allowedProtocolsString = allowedProtocolsString.replace(
            /,\s(?=[^,]*$)/,
            " or ",
          );
        }

        return `The URL must start with ${allowedProtocolsString}`;
      },
    _maxURLLength:
      ({ maxURLLength, label }) =>
      (value) => {
        if (!maxURLLength) return true;
        return (
          (value as string).length <= maxURLLength ||
          `${typeof label === "string" ? label : "URL"} must not exceed ${maxURLLength} characters`
        );
      },
  },
});

UrlField.displayName = "UrlField";

export { UrlField, type UrlFieldProps };
