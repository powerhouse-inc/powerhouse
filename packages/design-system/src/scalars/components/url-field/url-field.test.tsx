import { Form } from "#scalars";
import { render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { UrlField } from "./url-field.js";

describe("UrlField", () => {
  it("should match snapshot", () => {
    const { container } = renderWithForm(
      <UrlField
        name="test-url"
        label="Website URL"
        description="Enter your website URL"
        placeholder="https://example.com"
        required
        warnings={["URL may be unreachable"]}
        errors={["Invalid URL format"]}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with basic props", () => {
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        placeholder="https://example.com"
      />,
    );

    expect(screen.getByTestId("url-field")).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("https://example.com"),
    ).toBeInTheDocument();
  });

  it("should show required indicator when required", () => {
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        required
      />,
    );

    expect(screen.getByText("*")).toBeInTheDocument();
    expect(screen.getByTestId("url-field")).toHaveAttribute("required");
  });

  it("should show description when provided", () => {
    renderWithForm(
      <UrlField
        name="test-url"
        label="Website URL"
        description="Enter your website URL"
      />,
    );

    expect(screen.getByText("Enter your website URL")).toBeInTheDocument();
  });

  it("should show error message when provided", async () => {
    renderWithForm(
      <UrlField
        name="test-url"
        label="Website URL"
        errors={["Invalid URL format"]}
      />,
    );

    await waitFor(() =>
      expect(screen.getByText("Invalid URL format")).toBeInTheDocument(),
    );
  });

  it("should show warning message when provided", () => {
    renderWithForm(
      <UrlField
        name="test-url"
        label="Website URL"
        warnings={["URL may be unreachable"]}
      />,
    );

    expect(screen.getByText("URL may be unreachable")).toBeInTheDocument();
  });

  it("should be disabled when disabled prop is true", () => {
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        disabled
      />,
    );

    expect(screen.getByTestId("url-field")).toBeDisabled();
  });

  it("should validate URL format", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <>
        <UrlField data-testid="url-field" name="test-url" label="Website URL" />
        <button type="submit">Submit</button>
      </>,
    );

    const input = screen.getByTestId("url-field");
    await user.type(input, "not-a-url{enter}");

    expect(
      await screen.findByText("not-a-url must be a valid URL"),
    ).toBeInTheDocument();
  });

  it("should be invalid if URL is to long", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        maxURLLength={10}
      />,
    );

    const input = screen.getByTestId("url-field");
    await user.type(input, "https://example.com/{enter}"); // 20 characters (including protocol and domain)

    expect(input).toHaveAttribute("aria-invalid", "true");
    expect(
      await screen.findByText("Website URL must not exceed 10 characters"),
    ).toBeInTheDocument();
  });

  it("should show warning when the URL could be truncated", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <UrlField data-testid="url-field" name="test-url" label="Website URL" />,
    );

    const input = screen.getByTestId("url-field");
    await user.type(input, "https://example.com/test...");
    await user.tab(); // trigger blur

    expect(await screen.findByText("URL may be truncated")).toBeInTheDocument();
  });

  it("should not show warnings if the prop is set false", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        showWarnings={false}
      />,
    );

    const input = screen.getByTestId("url-field");
    await user.type(input, "https://example.com/test...");
    await user.tab(); // trigger blur

    expect(screen.queryByText("URL may be truncated")).not.toBeInTheDocument();
  });

  it("should be valid when valid", async () => {
    const user = userEvent.setup();
    const onSubmit = vi.fn();
    renderWithForm(
      <>
        <UrlField data-testid="url-field" name="test-url" label="Website URL" />
        <button type="submit">Submit</button>
      </>,
      onSubmit,
    );

    const input = screen.getByTestId("url-field");
    await user.type(input, "https://example.com{enter}");
    expect(onSubmit).toHaveBeenCalled();
  });

  it("should render platform built-in icon when URL matches configured hostname", () => {
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        value="https://github.com/test"
        platformIcons={{
          "github.com": "Github",
        }}
      />,
    );

    expect(screen.getByTestId("icon-fallback")).toBeInTheDocument();
  });

  it("should render a custom icon when hostname has no configured icon", () => {
    renderWithForm(
      <UrlField
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        value="https://github.com/test"
        platformIcons={{
          "github.com": <span data-testid="custom-icon">Custom Icon</span>,
        }}
      />,
    );

    expect(screen.getByTestId("custom-icon")).toBeInTheDocument();
  });

  it("should not render any icon when platformIcons prop is not provided", () => {
    renderWithForm(
      <UrlField data-testid="url-field" name="test-url" label="Website URL" />,
    );

    expect(screen.queryByTestId("icon-fallback")).not.toBeInTheDocument();
  });

  it("should hide warnings when the form is reset", async () => {
    const user = userEvent.setup();
    render(
      <Form onSubmit={() => undefined} defaultValues={{ "test-url": "" }}>
        {({ reset }) => (
          <>
            <UrlField
              data-testid="url-field"
              name="test-url"
              label="Website URL"
            />
            <button type="reset" onClick={reset} data-testid="reset-button">
              Reset
            </button>
          </>
        )}
      </Form>,
    );

    // first show the warnings typing in the input
    await user.type(screen.getByTestId("url-field"), "https://x.com/test...");
    await user.tab(); // trigger blur
    expect(await screen.findByText("URL may be truncated")).toBeInTheDocument();

    // then hide the warnings when the form is reset
    await user.click(screen.getByTestId("reset-button"));
    expect(screen.queryByText("URL may be truncated")).toBeNull();
  });
});
