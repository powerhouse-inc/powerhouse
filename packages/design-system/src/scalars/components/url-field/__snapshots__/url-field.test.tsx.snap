// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlField > should match snapshot 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-[22px] dark:text-gray-50 dark:group-hover:text-slate-50 text-red-700 group-hover:text-red-900"
        for=":r0:"
      >
        Website URL
        <span
          class="ml-1 text-blue-900 group-hover:text-gray-900 dark:group-hover:text-slate-50"
        >
          *
        </span>
      </label>
      <div
        class="flex items-center"
      >
        <input
          class="flex h-10 rounded-md text-sm text-gray-900 dark:text-gray-50 border border-gray-300 bg-white dark:border-charcoal-700 dark:bg-charcoal-900 px-3 py-2 font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 dark:focus-visible:ring-charcoal-300 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500 w-auto rounded-r-none border-r-0 focus:z-10"
          readonly=""
          style="width: 64px;"
          value="http://"
        />
        <input
          class="flex h-10 w-full rounded-md text-sm text-gray-900 dark:text-gray-50 border border-gray-300 bg-white dark:border-charcoal-700 dark:bg-charcoal-900 px-3 py-2 font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 dark:focus-visible:ring-charcoal-300 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500 focus:z-10 rounded-l-none"
          id=":r0:"
          name="test-url"
          placeholder="https://example.com"
          required=""
          type="url"
          value=""
        />
      </div>
      <p
        class="font-inter text-sm font-normal leading-5 text-gray-600"
      >
        Enter your website URL
      </p>
      <p
        class="mb-0 inline-flex items-center text-xs font-medium text-orange-900 dark:text-orange-900"
        data-type="warning"
      >
        URL may be unreachable
      </p>
      <p
        class="mb-0 inline-flex items-center text-xs font-medium text-red-900 dark:text-red-700"
        data-type="error"
      >
        Invalid URL format
      </p>
    </div>
  </form>
</div>
`;
